src/trace-plugin-loader.ts

```diff
@@ -15,26 +15,113 @@
  */
 'use strict';
 
+// TODO(kjin)
+// Module augmentation is implemented in this PR:
+// https://github.com/DefinitelyTyped/DefinitelyTyped/pull/19612
+// Do this correctly when it's landed and released.
+import _Module = require('module');
+const Module: {
+  _resolveFilename(request: string, parent?: NodeModule): string;
+  _load(request: string, parent?: NodeModule, isMain?: boolean): any;
+} = _Module as any;
+
+import { Logger } from '@google-cloud/common';
+import * as path from 'path';
+import * as semver from 'semver';
+import * as shimmer from 'shimmer';
 import * as util from './util';
 
-var Module = require('module');
-var shimmer = require('shimmer');
-var path = require('path');
-var semver = require('semver');
-var TraceAgent = require('./trace-api'/*.js*/);
+const TraceAgent = require('./trace-api');
+// TODO(kjin): When it's is converted, import these from trace-api
+export interface TraceAgentConfig {}
+export interface TraceAgent {
+  enable: (logger: Logger, config: TraceAgentConfig) => void;
+  disable: () => void;
+}
+
+/**
+ * An interface representing config options read by the plugin loader, which includes
+ * TraceAgent configuration as well.
+ */
+export interface PluginLoaderConfig extends TraceAgentConfig {
+  plugins: {
+    [pluginName: string]: string;
+  };
+}
+interface InternalPatch<T> {
+  file: string;
+  module?: T;
+  patch: (module: T, agent: TraceAgent) => void;
+  unpatch?: (module: T) => void;
+}
+
+interface InternalIntercept<T> {
+  file: string;
+  module?: T;
+  intercept: <T>(module: T, agent: TraceAgent) => T;
+}
+
+interface InternalPlugin {
+  file: string;
+  patches: {
+    [patchName: string]: {
+      [file: string]: InternalPatch<any> | InternalIntercept<any>;
+    }
+  };
+  agent: TraceAgent;
+}
+
+interface PluginStore {
+  [pluginName: string]: InternalPlugin;
+}
+
+export interface Patch {
+  file?: string;
```

Comment:
nit: `file` is not optional in `InternalPatch`. Is this intentional?

Reply:
Yes, omitting `file` is equivalent to specifying that the export of the module itself should be patched (from the [plugin guide](https://github.com/GoogleCloudPlatform/cloud-trace-nodejs/blob/master/doc/plugin-guide.md))

Reply:
My concern was around the fact that for internal patches we require that the `file` must always be specified. If we already have the code (and tests) to deal with optional `file` field, do we really get anything out of mandating `file` for internal patches?

Reply:
Put differently, I want to make sure there is value we get out of having different types for `InternalPatch` and `Patch`.

Reply:
It's true that data store in an `InternalPatch` is derived mostly from `Patch`, but we already want to hold other information, such as `module`, a pointer to the loaded module being patched. Since this other information is purely implementation detail, it wouldn't be appropriate to expose it to the user (`Patch` is a user-facing interface).

Given that there must be a distinction between `InternalPatch` and `Patch`, I don't see why we should make `InternalPatch::file` optional. In my mind, any internal interface should be as strictly-defined as possible, while anything user-facing should be as loose as needed -- this is a principle that we already apply to the agent configuration object in `index.ts::initConfig`: to the user, every configuration field appears to be optional, but internally, we copy the user-set fields into an object with a more strictly-defined shape, substituting default values for user-omitted fields.

Reply:
The one reason I can think of is that in the current structure they are unrelated interfaces. I would prefer one to be a subtype of the other. That leads to more code sharing.

Reply:
Between internal and external representations of some data:

1. The external representation should be at least more permissive than the internal representation, allowing users to omit fields.  The implementation will populate the internal representation with default values (`InternalPatch::file`)
2. The internal representation may have extra fields that are not relevant in the external representation (`InternalPatch::module`)
3. One should be a subtype of the other, or they should otherwise share common code in some way

My belief is that these points are generally applicable, beyond just interface definitions. For example, the first two points apply to functions as well: functions accept optional arguments, initialize them to default values if omitted, and also have function-scope "implementation detail" variables.

These three things are all nice to have (feel free to dispute that though), but AFAIK in TypeScript we can only pick two, because in (1) the internal representation is a subtype of the external representation, and vice-versa in (2). Given this, I would prefer to have the first two despite the slightly improved code brevity provided by (3), because the first two have to do with code correctness, and the third code readability, which is arguably less important.

Maybe a different way of putting this is -- between a public-facing and implementation-specific representation, which should be a subtype of the other?

Reply:
Can't we get all three? This is valid TS.

```ts
interface Patch {
  file?: string;
}

interface InternalPatch extends Patch {
  file: string;
}

function foo(p: Patch) {}

const q: InternalPatch = { file: 'foo' };

foo(q);
```

Reducing code duplication doesn't just give us brevity, it is also leads to fewer opportunities for bugs or for code to incidentally diverge.

Reply:
Oops, you're right